### PR TITLE
storage-mon: new resource agent

### PR DESCRIPTION
**Pull request by fabbione**
Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

### CODE REVIEW

**Comment by fabbione** [comment]
@oalbrigt the ras should be ok, the C code needs some tuning. Can you please start review the changes?

File: doc/man/Makefile.am [hunk 1]
```diff
@@ -183,7 +183,8 @@ man_MANS                = ocf_heartbeat_AoEtarget.7 \
                           ocf_heartbeat_vdo-vol.7 \
                           ocf_heartbeat_vmware.7 \
                           ocf_heartbeat_vsftpd.7 \
-                          ocf_heartbeat_zabbixserver.7
+                          ocf_heartbeat_zabbixserver.7 \
+                          ocf_heartbeat_storage-mon.7
```
Comment (oalbrigt) [inline doc/man/Makefile.am:187]: I prefer to keep these sorted.

File: configure.ac [hunk 2]
```diff
@@ -1006,6 +1006,7 @@ AC_CONFIG_FILES([heartbeat/sybaseASE], [chmod +x heartbeat/sybaseASE])
 AC_CONFIG_FILES([heartbeat/syslog-ng], [chmod +x heartbeat/syslog-ng])
 AC_CONFIG_FILES([heartbeat/vsftpd], [chmod +x heartbeat/vsftpd])
 AC_CONFIG_FILES([heartbeat/CTDB], [chmod +x heartbeat/CTDB])
+AC_CONFIG_FILES([heartbeat/storage-mon], [chmod +x heartbeat/storage-mon])
```
Comment (oalbrigt) [inline configure.ac:1009]: These should be sorted the way the rest of the list are as well.

File: heartbeat/storage-mon.in [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+: ${OCF_FUNCTIONS_DIR=${OCF_ROOT}/lib/heartbeat}
+. ${OCF_FUNCTIONS_DIR}/ocf-shellfuncs
+: ${__OCF_ACTION:="$1"}
+
+#
+STORAGEMON=$HA_BIN/storage_mon
+ATTRDUP=/usr/sbin/attrd_updater
+
+# Explicitly list all environment variables used, to make static analysis happy
+: ${OCF_RESKEY_CRM_meta_interval:=0}
```
Comment (oalbrigt) [inline heartbeat/storage-mon.in:56]: These should be set in _default parameters to be used in metadata (so metadata doesnt change if you provided some parameters)

Example:
https://github.com/ClusterLabs/resource-agents/blob/master/heartbeat/IPaddr2#L71

File: heartbeat/storage-mon.in [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+System health agent that checks the storage I/O status of the given drives and
+updates the #health-storage attribute. Usage is highly recommended in combination
+with storage-mon monitoring agent. The agent currently support a maximum of 25
+devices per instance.
+</longdesc>
+<shortdesc lang="en">storage I/O health status</shortdesc>
+
+<parameters>
+
+<parameter name="state" unique="1">
```
Comment (oalbrigt) [inline heartbeat/storage-mon.in:81]: state_file is probably a better name, so you know what to enter/expect at first glance.

File: heartbeat/storage-mon.in [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+	if [ ! -x "$STORAGEMON" ] ; then
+		ocf_log err "${STORAGEMON} not installed."
+		exit $OCF_ERR_INSTALLED
+	fi
+
+	i=0
+	for DRIVE in ${OCF_RESKEY_drives}; do
+		if [ ! -e "$DRIVE" ] ; then
+			ocf_log err "${DRIVE} not found on the system"
+			exit $OCF_ERR_INSTALLED
```
Comment (oalbrigt) [inline heartbeat/storage-mon.in:149]: I think _CONFIGURED is better for this and the rest of the returns in this function..

File: heartbeat/storage-mon.in [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+}
+
+storage-mon_validate() {
+	storage-mon_init
+
+	# Is the state directory writable?
+	state_dir=$(dirname "$OCF_RESKEY_state")
+	touch "$state_dir/$$"
+	if [ $? -ne 0 ]; then
+		return $OCF_ERR_ARGS
```
Comment (oalbrigt) [inline heartbeat/storage-mon.in:179]: This should probably be _CONFIGURED.

File: heartbeat/storage-mon.in [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+}
+
+storage-mon_validate() {
+	storage-mon_init
+
+	# Is the state directory writable?
+	state_dir=$(dirname "${OCF_RESKEY_state}")
+	touch "$state_dir/$$"
+	if [ $? -ne 0 ]; then
+		return $OCF_ERR_ARGS
```
Comment (oalbrigt) [inline heartbeat/storage-mon.in:240]: This should probably be _CONFIGURED.

**Comment by fabbione** [comment]
@oalbrigt should all be done now

File: heartbeat/storage-mon.in [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+: ${OCF_FUNCTIONS_DIR=${OCF_ROOT}/lib/heartbeat}
+. ${OCF_FUNCTIONS_DIR}/ocf-shellfuncs
+: ${__OCF_ACTION:="$1"}
+
+#
+STORAGEMON=$HA_BIN/storage_mon
+ATTRDUP=/usr/sbin/attrd_updater
+
+OCF_RESKEY_CRM_meta_interval_default="0"
+OCF_RESKEY_drives_default="/dev/sda"
```
Comment (oalbrigt) [inline heartbeat/storage-mon.in:56]: I think it's better to not have default for drives, and have it required, as it's likely to not be /dev/sda in 50%+ of setups due to LVM/dm.

**Comment by fabbione** [comment]
@oalbrigt should be fixed now

**Comment by oalbrigt** [comment]
Thanks.

**Pull request merged**